snuba/query/data_source/simple.py

```diff
@@ -32,3 +34,21 @@ class Entity(SimpleDataSource):
 
     def get_columns(self) -> ColumnSet:
         return self.schema
+
+
+@dataclass(frozen=True)
+class Table(SimpleDataSource):
+    """
+    Represents a table or a view in the physical query.
+    """
+
+    table_name: str
+    schema: ColumnSet
+    # TODO: Move mandatory connditions ad prewhere candidates out of
+    # here as they are structural property of a storage. This requires
+    # the processors that consume these fields to access the storage.
+    mandatory_conditions: Sequence[FunctionCall] = field(default_factory=list)
+    prewhere_candidates: Sequence[str] = field(default_factory=list)
+
+    def get_columns(self) -> ColumnSet:
+        return self.schema
```

Comment:
Having both Entity and Table modelled as types of SimpleDataSource seems a bit uncanny to me - they don't appear to have too much in common. Well, Table as a data source is more intuitive than Entity, since Entity does not seem like a data source at all (the actual data source of an entity can potentially be picked from multiple tables, or I guess potentially even a join or subquery or non ClickHouse source). Should the logical query even have a "data source" or "from clause", or should it instead just hold the entity or entity relationship as it is requested with Snql, which can be later transformed to a data source in the ClickhouseQuery?

Reply:
> Should the logical query even have a "data source" or "from clause", or should it instead just hold the entity or entity relationship as it is requested with Snql, which can be later transformed to a data source in the ClickhouseQuery?

It actually does both.
A simple logical query has a from_clause that is an Entity. Simple Query on one entity.
A composite query can have a from_clause that is either a join or a subquery. Each subquery is either composite or is a simple query with a from clause that is an individual Entity.

I am not sure I get exactly the suggestion, so just to be sure we are onn the same page:
DataSource is the from clause of a query basically. So if you look at the data source hierarchy, it contains Entity, Table, Query, Join, and CompositeQuery (which is a QUery). 
These are requirements:
- both the logical and the clickhouse queries need to have a from_clause (the data source represents that).
- in both the logical and clickhouse queries the data source can be either simple (entity in one case, table in the other) or complex (join or subquery). This is needed for query processing.

Now this design uses the same abstract data source because there are plenty of commonailities between the logical and clickhouse data sources for what concerns subqueries and joins. In those cases the differences are minute. So I made them Generic with respect to the simple data source type: a logical query is a Query[Entity] and a clickhouse query is a Query[Table]. While the join is literally the same class: Join[Entity] and Join[Table].

If it was not for the commonalities to deal with for joins and subqueries (which is a lot of logic) there would have been no reason to have the same data source hierarchy.

Does this make sense?

Reply:
Yeah, I see. What about adding LogicalDataSource and PhysicalDataSource subclasses so that we can ensure the correct types are present in the logical and physical query since they are not supposed to be interchangeable?

Reply:
We could not because CompositeQuery and Join are data sources and the class itself is supposed to be used in both cases.
Though we achieve the result of making them not interchangeable by making them generic:
https://github.com/getsentry/snuba/blob/master/snuba/query/composite.py#L17

`CompositeQuery[Entity]` is a logical data source
`CompositeQuery[Table]` is a physical data source

It works the same way for joins.

Essentially all the data sources that are not strictly logical or physical (Entity and Table are strictly logical and physical) are generic with respect to a simple tada source which makes then univocally logical or physical. 
So all the code of the logical query processing depends on Entity or data sources that are Generic[Entity] while all the physical query processing depends on Table or data sources that are Generic[Table]

Reply:
Right, makes sense